### PR TITLE
Editor controls template labels

### PR DESCRIPTION
**Pull request by mattben**

### CODE REVIEW

File: frontend/src/js/components/modals/TemplateEditor.jsx [hunk 1]
```diff
@@ -1,14 +1,21 @@
...
+        template: React.PropTypes.object,
+        updateTemplate: React.PropTypes.func,
+    },
+    onUpdate(attribute, key, value) {
+        this.props.updateTemplate(this.props.template.name, attribute, key, value);
     },
     render() {
         let template = this.props.template;
         let template_name = template ? template.name : "";
+        console.log("Template name", template_name)
```
Comment (chaosphere2112) [inline frontend/src/js/components/modals/TemplateEditor.jsx]: Please remove console.log statements

File: frontend/src/js/containers/LeftSideBar.jsx [hunk 2]
```diff
@@ -47,6 +48,7 @@ var LeftSideBar = React.createClass({
         this.initDragListItems();
     },
     render() {
+        console.log("lsb", this.props.updateTemplate);
```
Comment (chaosphere2112) [inline frontend/src/js/containers/LeftSideBar.jsx]: Please remove console.log statements

File: frontend/src/js/components/TemplatePreview.jsx [hunk 3]
```diff
@@ -7,42 +7,17 @@ var TemplatePreview = React.createClass({
         template: React.PropTypes.object
     },
     componentDidUpdate(prevProps, prevState) {
-        if (this.props.template) {
```
Comment (chaosphere2112) [inline frontend/src/js/components/TemplatePreview.jsx]: Why was this logic removed? Just curious.

Comment (mattben) [inline frontend/src/js/components/TemplatePreview.jsx]: explained in office.

File: frontend/src/js/components/editors/template/TemplateLabelSettings.jsx [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+
+var TemplateLabelSettings = React.createClass({
+    propTypes: {
+        label: React.PropTypes.object,
+        update: React.PropTypes.func
+    },
+    updateLabelValue(key) {
+        const self = this;
+        let validator = parseFloat;
```
Comment (chaosphere2112) [inline frontend/src/js/components/editors/template/TemplateLabelSettings.jsx]: Can we also validate that it's between 0 and 1 if it's a float? Anything greater than 1 is invalid and will make things complain.

File: frontend/src/js/components/modals/TemplateEditor.jsx [hunk 5]
```diff
@@ -1,10 +1,15 @@
...
 
 
 var TemplateEditor = React.createClass({
     propTypes: {
-        template: React.PropTypes.object
+        template: React.PropTypes.object,
+        updateTemplate: React.PropTypes.func,
+    },
+    onUpdate(attribute, key, value) {
+        this.props.updateTemplate(this.props.template.name, attribute, key, value);
```
Comment (chaosphere2112) [inline frontend/src/js/components/modals/TemplateEditor.jsx]: We should update this to use the React `state` system to have an internal template that we're editing, and then a "Save" button that syncs the internal template with the external one (so we can have a "cancel" button that resets everything appropriately by *not* syncing and just killing the modal).

The code you need to add looks something like this:

```
getInitialState() {
    return {workingTemplate: $.extend({}, this.props.template)};
},
componentWillReceiveProps(nextProps) {
    this.setState({workingTemplate: $.extend({}, nextProps.template)});
}
```

and then using `this.state.workingTemplate` where we're using `this.props.template` currently.

You can come to me with questions about that.

Comment (chaosphere2112) [inline frontend/src/js/components/modals/TemplateEditor.jsx]: Note that you only need to do that in this file, and everything that happens internally to the subcomponents should be able to stay the same.

File: frontend/src/js/components/editors/template/TemplateLabelSettings.jsx (hunk 4 above)
Comment (mattben) [inline frontend/src/js/components/editors/template/TemplateLabelSettings.jsx]: What is validator? is it a function? `e.tartget.value` if less than 0 then equal 0; if greater than 1 then equal 1;

File: frontend/src/js/components/editors/template/TemplateLabelSettings.jsx [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    },
+    range(alt){
+        if(alt < 0){
+            alt = 0;
+        }
+        else if (alt > 1){
+            alt = 1;
+        }
+        return alt;
+    },
```
Comment (mattben) [inline frontend/src/js/components/editors/template/TemplateLabelSettings.jsx]: So this works, and is ugly but we still can't enter a decimal number.

if you type some thing target then one its cast back to one, if you type a decimal point `.` I throws `ValueError: y must be a number`  

so to type `.5` I had to type 1 back arrow add the dot `.1` add the 5 `.15` then remove the 1 `.5` ... thats not right... 

Thoughts?

File: frontend/src/js/components/editors/template/TemplateLabelsEditor.jsx [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    },
+    getInitialState() {
+        return {workingTemplate: $.extend({}, this.props.template)};
+    },
+    componentWillReceiveProps(nextProps) {
+        this.setState({workingTemplate: $.extend({}, nextProps.template)});
+    },
+    render() {
+        const template = this.state.workingTemplate;
+        const self = this;
```
Comment (mattben) [inline frontend/src/js/components/editors/template/TemplateLabelsEditor.jsx]: your code additions seem to work fine now how do you want me to wire up the save and cancel buttons (see below)

Comment (mattben) [inline frontend/src/js/components/editors/template/TemplateLabelsEditor.jsx]: opps is this the wrong file? Id doesn't break anything

File: frontend/src/js/components/modals/TemplateEditor.jsx [hunk 8]
```diff
@@ -22,6 +27,9 @@ var TemplateEditor = React.createClass({
                             </h4>
                         </div>
                         <TemplatePreview template={template} />
+                        <TemplateLabelsEditor template={template} updateTemplate={this.onUpdate}/>
+                        <button type="button" aria-label="Save">Save</button>
+                        <button type="button" aria-label="Cancel">Cancel</button>
                     </div>
```
Comment (mattben) [inline frontend/src/js/components/modals/TemplateEditor.jsx]: save needs to save state `this.props.template = this.state.workingTemplate`
cancel needs to revert and close 'this.state.workingTemplate = his.props.template` && close

Comment (mattben) [inline frontend/src/js/components/modals/TemplateEditor.jsx]: save needs to save state this.props.template = this.state.workingTemplate
cancel needs to revert and close 'this.state.workingTemplate = his.props.template` && close

File: frontend/src/js/components/editors/template/TemplateLabelSettings.jsx (hunk 4 above)
Comment (chaosphere2112) [inline frontend/src/js/components/editors/template/TemplateLabelSettings.jsx]: Yeah, `validator` is a function.

File: frontend/src/js/components/editors/template/TemplateLabelSettings.jsx (hunk 6 above)
Comment (chaosphere2112) [inline frontend/src/js/components/editors/template/TemplateLabelSettings.jsx]: Yeah, the general approach is "right", but we need to change the specifics. Basically, you want to use the onChange handler to "validate" the data and make sure it's good (integer for priority,  0<= float <=1 for others); you store the string value in this.state (using `this.setState({"key": "value"})`), and have a separate state value of "valid" that's a boolean; if a field is invalid, you stick a class on the field that will mark it as so. Then you can use `onBlur` to actually sync the value out to the parent.

File: frontend/src/js/components/editors/template/TemplateLabelSettings.jsx [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    render() {
+        const x = this.props.label.x;
```
Comment (chaosphere2112) [inline frontend/src/js/components/editors/template/TemplateLabelSettings.jsx]: You'll want to store these values in the component's `state`.

File: frontend/src/js/components/editors/template/TemplateLabelSettings.jsx [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+        const y = this.props.label.y;
+        const priority = this.props.label.priority;
+        const name = this.props.label.member;
+
+        return (
+            <tr>
+                <th>{name}</th>
+                <td><input type="number" value={x} name="{name}_x" onChange={this.updateLabelValue("x")} /></td>
```
Comment (chaosphere2112) [inline frontend/src/js/components/editors/template/TemplateLabelSettings.jsx]: you'll want to move updateLabelValue to `onBlur` and move the validation logic out of updateLabelValue to a new `onChange` function that updates the component's state.

File: frontend/src/js/components/editors/template/TemplateLabelsEditor.jsx (hunk 7 above)
Comment (chaosphere2112) [inline frontend/src/js/components/editors/template/TemplateLabelsEditor.jsx]: This is the wrong file to handle the state; you should move the workingTemplate stuff to the TemplateEditor. Then you can wire the save/cancel buttons to sync the prop or not at that level, rather then having to pass stuff around.

**Comment by chaosphere2112** [comment]
@mattben Provided input; let me know if you have more questions.

**Pull request closed without merging**